nemo/collections/nlp/models/question_answering/qa_model.py

```diff
@@ -178,51 +172,28 @@ def validation_epoch_end(self, outputs):
+        start_logits = s.squeeze().cpu().numpy().tolist()
+        end_logits = e.squeeze().cpu().numpy().tolist()
+        (all_predictions, all_nbest, scores_diff) = self.test_dataset.get_predictions(
+            unique_ids=unique_ids,
+            start_logits=start_logits,
+            end_logits=end_logits,
+            n_best_size=self._cfg.test_ds.n_best_size,
+            max_answer_length=self._cfg.test_ds.max_answer_length,
+            version_2_with_negative=self._cfg.dataset.version_2_with_negative,
+            null_score_diff_threshold=self._cfg.test_ds.null_score_diff_threshold,
+            do_lower_case=self._cfg.dataset.do_lower_case,
+        )
 
-        all_unique_ids = []
-        all_logits = []
-        if torch.distributed.is_initialized():
-            world_size = torch.distributed.get_world_size()
-            for ind in range(world_size):
-                all_unique_ids.append(torch.empty_like(unique_ids))
-                all_logits.append(torch.empty_like(logits))
-            torch.distributed.all_gather(all_unique_ids, unique_ids)
-            torch.distributed.all_gather(all_logits, logits)
-        else:
-            all_unique_ids.append(unique_ids)
-            all_logits.append(logits)
-
-        if not torch.distributed.is_initialized() or torch.distributed.get_rank() == 0:
-
-            unique_ids = []
-            start_logits = []
-            end_logits = []
-            for u in all_unique_ids:
-                unique_ids.extend(u.cpu().numpy().tolist())
-            for u in all_logits:
-                s, e = u.split(dim=-1, split_size=1)
-                start_logits.extend(s.squeeze().cpu().numpy().tolist())
-                end_logits.extend(e.squeeze().cpu().numpy().tolist())
-
-            (all_predictions, all_nbest, scores_diff) = self.test_dataset.get_predictions(
-                unique_ids=unique_ids,
-                start_logits=start_logits,
-                end_logits=end_logits,
-                n_best_size=self.test_config.n_best_size,
-                max_answer_length=self.test_config.max_answer_length,
-                version_2_with_negative=self.version_2_with_negative,
-                null_score_diff_threshold=self.test_config.null_score_diff_threshold,
-                do_lower_case=self.do_lower_case,
-            )
-
-            if self.test_config.output_nbest_file is not None:
-                with open(self.test_config.output_nbest_file, "w") as writer:
-                    writer.write(json.dumps(all_nbest, indent=4) + "\n")
-            if self.test_config.output_prediction_file is not None:
-                with open(self.test_config.output_prediction_file, "w") as writer:
-                    writer.write(json.dumps(all_predictions, indent=4) + "\n")
+        if self._cfg.test_ds.output_nbest_file is not None:
+            with open(self._cfg.test_ds.output_nbest_file, "w") as writer:
+                writer.write(json.dumps(all_nbest, indent=4) + "\n")
```

Comment:
check it's done only on rank 0?

Reply:
shouldn't test() be run on a single gpu anyway? On multi-gpu the result could  be incorrect

Reply:
shouldn't test() be run on a single GPU anyway?

Reply:
ok added check to script, thanks!